src/mlpack/methods/ann/vanilla_rbm_impl.hpp

```diff
@@ -0,0 +1,201 @@
+           HiddenLayerType>::Train(const arma::mat& predictors, 
+                                   Optimizer<NetworkType>& optimizer)
+{
+
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+  {
+    Reset();
+  }
+
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+
+ /* 
+  * This function calculates
+  * the free energy of the model
+  */
+template<typename InitializationRuleType,
+         typename VisibleLayerType,
+         typename HiddenLayerType>
+double RBM<InitializationRuleType, 
+           VisibleLayerType, 
+           HiddenLayerType>::FreeEnergy(const arma::mat&& input)
+{
+  arma::mat output;
+  visible.Forward(std::move(input), std::move(output));
+  SoftplusFunction::Fn(output, output);
+  return  -arma::dot(input, visible.Bias()) - arma::accu(output);
+}
+
+template<typename InitializationRuleType,
+  typename VisibleLayerType,
+  typename HiddenLayerType>
+double RBM<InitializationRuleType, 
+  VisibleLayerType, 
+  HiddenLayerType>::Evaluate(const arma::mat& /* parameters*/, arma::vec& orderFunction)
+{
+  double freeEnergy, freeEnergyCorrupted;
+  arma::Col<double> output(orderFunction.n_elem);
+  size_t corruptIdx = math::RandInt(0, visible.Bias().n_elem);
+  for(auto i: orderFunction)
+  {
+    // Do not use unsafe col predictor as we change the input
+    arma::vec currentInput = predictors.col(i);
+    freeEnergy = FreeEnergy(std::move(currentInput)); 
+    currentInput(corruptIdx) =  1 - predictors(corruptIdx);
+    freeEnergyCorrupted = FreeEnergy(std::move(currentInput));
+    output(i) = SoftplusFunction::Fn(freeEnergyCorrupted - freeEnergy + 1);
+  }
+  // Concerned here for neumerical stability.
+  return arma::accu(output) * visible.Bias().n_elem;
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType, typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, 
+  HiddenLayerType>::SampleHidden(arma::mat&& input, arma::mat&& output)
+{
+  visible.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType, typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, 
+  HiddenLayerType>::SampleVisible(arma::mat&& input, arma::mat&& output)
+{
+  hidden.Sample(std::move(input), std::move(output));
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType, typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, 
+  HiddenLayerType>::Gibbs(arma::mat&& input, arma::mat&& negative_sample, size_t num_steps , bool persistence)
+{
+  arma::mat temp, temp1;
+  temp = input;
+
+  if(persistence && !state.is_empty())
+    input = state;
+  
+  for(size_t j = 0; j < num_steps - 1; j++)
+  {
+    // Use probabilties for updation till the last step(section 3 hinton)
+    ForwardVisible(std::move(temp), std::move(temp1));
```

Comment:
I think you can read about it here. https://www.cs.toronto.edu/~hinton/absps/guideTR.pdf.
But earlier i was doing just sampleVisible and sampleHidden for numSteps then also the result was not converging